### PR TITLE
Refactors and enhancements from my develop branch.

### PR DESCRIPTION
**Pull request by topisani**
While working on #212 and integrating kts in my workflow, I made some refactors and improvements on my own develop branch. I would like to keep working on kts, but I also don't want to further fracture the efforts, so i hope we can find a way to merge some of this.

Other than what is already in #212, this includes:

## Features:
 - Get defined faces from kakoune automatically using `debug faces`. This works, and removes the need of defining the available highlight groups in `config.toml`, but it has a few drawbacks:
   - currently, it isnt run automatically. There is no hook for changing colorscheme, so its hard to find the correct hook. I have it run once on startup in my own setup.
   - It parses the `*debug*` buffer, which works fine, but means that it fills lots of lines into the debug buffer every time it runs. I would like to contribute something to kakoune that improves this, possibly just a `debug -to-file`, or similar.
 - builds on top of #212 to make textobjects multiple selections compatible in a performant way.

## Enhancements
 - Only request highlighting if the buffer has changed (based on timestamp)
 - Improved handling of crashed/stopped servers by deleting the fifos when the server stops, and handling non-existing fifos when sending requests. When a new server is started, the clients reconnect and reinitialize automatically.
 - Minor: Add documentation comment highlighting to cpp files (compatible with existing faces)

## Refactoring
 - Split `server.rs` out into a few submodules
 - Move the handling of individual commands from `server` to `handler`
 - Use a `RequestMeta` type to simplify handling common request data (client, buffer, lang, timestamp)
 - Split `static.kak` into multiple files
 - Add `kak-tree-sitter-send-cmd` and `kak-tree-sitter-send-buf` helpers in `base.kak`
 - Use a shell block instead of kakoune options to define the colors for the default theme.
 - Use `%exp{}` blocks for request formatting instead of escaping in `"` strings.

All kakscript still uses shell blocks as little as possible (no shell blocks were added).

### CODE REVIEW

**Comment by phaazon** [comment]
Thanks! Should we close #212 and superseded by this current one?

File: kak-tree-sitter/rc/features.kak [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+      kak-tree-sitter-send-cmd %exp{{ "type": "highlight" }}
+      kak-tree-sitter-send-buf
+      set-option buffer kts_highlight_timestamp %val{timestamp}
+    }
+  }
+}
+
+# Enable highlighting for the current buffer.
+# 
+# This command does a couple of things, among removing the « default » highlighting (Kakoune based) of the buffer and
```
Comment (phaazon) [inline kak-tree-sitter/rc/features.kak:30]: I’m not sure where you remove it?

File: kak-tree-sitter/src/error.rs [hunk 2]
```diff
@@ -68,6 +70,18 @@ pub enum OhNo {
   #[error("cannot send request: {err}")]
   CannotSendRequest { err: String },
 
+  #[error("cannot parse buffer")]
+  CannotParseBuffer,
```
Comment (phaazon) [inline kak-tree-sitter/src/error.rs:74]: Could be worth it to add the name of the buffer in the error.

File: kak-tree-sitter/src/error.rs [hunk 3]
```diff
@@ -68,6 +70,18 @@ pub enum OhNo {
...
+
   #[error("highlight error: {err}")]
   HighlightError { err: String },
+  
+  #[error("textobject error: {err}")]
+  TextobjectError { err: String },
+  
+  #[error("internal server error: {err}")]
+  InternalError { err: String },
```
Comment (phaazon) [inline kak-tree-sitter/src/error.rs:83]: Still not a fan of such a vague error type. 🙂

File: kak-tree-sitter/src/handler.rs [hunk 4]
```diff
@@ -1,33 +1,107 @@
...
+    session_name: &str,
+    meta: &RequestMeta,
+    cmd: &SessionCmd,
+    buf: Option<&str>,
+  ) -> Result<Option<Response>, OhNo> {
+    match cmd {
+      SessionCmd::TryEnableHighlight => self
+        .handle_try_enable_highlight(session_name, &meta.lang)
+        .map(Some),
+      SessionCmd::DeclareFaces { faces } => {
```
Comment (phaazon) [inline kak-tree-sitter/src/handler.rs:68]: I like that idea; it should easy everything.

File: kak-tree-sitter/src/handler.rs [hunk 5]
```diff
@@ -58,34 +132,141 @@ impl Handler {
...
-    timestamp: u64,
+    meta: &RequestMeta,
     buf: &str,
   ) -> Result<Response, OhNo> {
     log::debug!(
-      "highlight for session {session_name}, buffer {buffer}, lang {lang_name}, timestamp {timestamp}"
+      "highlight for session {session_name}, buffer {}, lang {}, timestamp {}",
+      meta.buffer,
+      meta.lang,
+      meta.timestamp
```
Comment (phaazon) [inline kak-tree-sitter/src/handler.rs:142]: For a debug log, I think you can just `{meta:#?}`; it’s fine.

File: kak-tree-sitter/src/request.rs [hunk 6]
```diff
@@ -4,10 +4,12 @@ use std::fmt::Debug;
 
 use serde::{Deserialize, Serialize};
 
+use crate::kak;
+
 /// Unidentified request (i.e. not linked to a given session).
 #[derive(Debug, Deserialize, Serialize)]
 #[serde(tag = "type", rename_all = "snake_case")]
-pub enum UnixRequest {
+pub enum GlobalRequest {
```
Comment (phaazon) [inline kak-tree-sitter/src/request.rs:12]: I’m not sure why you renamed it `GlobalRequest`?

File: kak-tree-sitter/src/response.rs [hunk 7]
```diff
@@ -100,6 +107,15 @@ impl Response {
           range_specs = "set buffer kts_highlighter_ranges",
         )
       }
+
+      Response::TextObjects {
+        timestamp,
+        obj_type: _,
+        ranges,
```
Comment (phaazon) [inline kak-tree-sitter/src/response.rs:114]:
```suggestion
        timestamp,
        ranges,
        ..
```

File: kak-tree-sitter/src/server/mod.rs [hunk 8]
```diff
@@ -0,0 +1,394 @@
+mod fifo;
```
Comment (phaazon) [inline kak-tree-sitter/src/server/mod.rs:1]: Please let that in `server.rs`.

File: kak-tree-sitter/src/server/mod.rs [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+
+          _ => (),
+        }
+      }
+    }
+
+    log::info!("shutting down");
+    // NOTE: If we disconnect the sessions, they dont automatically reconnect.
+    // If we dont, the next request from a session automatically starts kak-tree-sitter again.
+    // self.disconnect_sessions();
```
Comment (phaazon) [inline kak-tree-sitter/src/server/mod.rs:318]: What do you mean with this?

File: kak-tree-sitter/src/server/runtime_dir.rs [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+      .ok_or_else(|| OhNo::NoRuntimeDir)?;
+
+    let base_dir = dir.join("kak-tree-sitter");
+    Ok(Self {
+      base_dir,
+      created: false,
+    })
+  }
+
+  pub fn display(&self) -> std::path::Display<'_> {
```
Comment (phaazon) [inline kak-tree-sitter/src/server/runtime_dir.rs:29]:
```suggestion
  pub fn display(&self) -> impl '_ + Display {
```

File: kak-tree-sitter/src/server/socket.rs [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+
+    Ok(Self {
+      is_standalone,
+      resources,
+      unix_listener,
+      resp_sender,
+    })
+  }
+
+  /// Register the Unix handler to a Poll.
```
Comment (phaazon) [inline kak-tree-sitter/src/server/socket.rs:57]: You named that socket handler; it’s not unix handler anymore.

File: runtime/queries/cpp/highlights.scm [hunk 12]
```diff
@@ -269,4 +269,10 @@
 
 (identifier) @variable
 
+((comment) @comment.documentation
+    (#match? @comment.documentation "^///\\s.*"))
+    
+((comment) @comment.documentation
+    (#match? @comment.documentation "^/\\*\\*\\s+.*"))
+
```
Comment (phaazon) [inline runtime/queries/cpp/highlights.scm:277]: What are those?

**Review by phaazon — CHANGES_REQUESTED** [review]
It’s an interesting PR that I wish we could merge ASAP. However, a couple of things:

- I left a bunch of suggestions of things to change.
- Some questions / nitpicking.
- You renamed / moved a lot of things and I struggle to understand why. I think it’s not needed _for now_ and you should probably focus on implementing a feature. You moved a lot of symbols that have nothing to do with textobjects, making reviewing longer and harder, and breaking a bit history search (i.e. I very often `git blame` inside my editor, and seeing the same commit all over the place is not good).

**Comment by topisani** [comment]
> - You renamed / moved a lot of things and I struggle to understand why. I think it’s not needed _for now_ and you should probably focus on implementing a feature. You moved a lot of symbols that have nothing to do with textobjects, making reviewing longer and harder, and breaking a bit history search (i.e. I very often `git blame` inside my editor, and seeing the same commit all over the place is not good).

I didn't originally do that with an intent to merge it, I was mainly just playing around with the codebase for my own use. I then ended up implementing some features on top of it. I should probably have marked this PR as a draft.

I'll try to split it out into some separate feature commits without the unnecessary refactorings.

File: kak-tree-sitter/rc/base.kak [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+  set -remove buffer %arg{1} %val{timestamp}
+  try %{
+    # This only succeeds if it finds the above function, i.e. if the value of the option in %arg{1} is now 0
+    eval "eval ""kak-tree-sitter--nop-exists-if-zero-%%opt{%arg{1}}"""
+    set buffer %arg{1} %val{timestamp}
+  } catch %{
+    set buffer %arg{1} %val{timestamp}
+    eval %arg{2}
+  }
+}
```
[the remark this answers is not in the document]
Comment (topisani) [reply kak-tree-sitter/rc/base.kak:159]: To only request highlighting when the timestamp has changed. Without using this, we would have to use a shell block to check the timestamp, which would start a new process on every NormalIdle.
I know it's very hacky, but It solves the problem pretty well, and I see no reason why it shouldn't continue to work reliably

File: runtime/queries/cpp/highlights.scm (hunk 12 above)
Comment (topisani) [reply runtime/queries/cpp/highlights.scm:277]: Documentation comments in cpp. Like in rust, you can do `/// ` to make a documentation comment, that's then picked up by systems like doxygen. `/** */` has the same effect

**Pull request closed without merging**